### DISCUSSION TITLE
capnproto for ipc

### DISCUSSION THREAD

**Discussion by elrom** [capnproto/capnproto#1993] [General]
Hi Kenton!

I'm interested in using cnp for shmem ipc. I'm really after zero copy etc. My use case is very resource constrained. 

I did some searching before posting,  and saw you mentioned this as part of rationale for having created cnp (I also saw it on the roadmap).

If I wanted to try to do this, what's the easiest path? Do I create my own custom allocator, so I can just hand cnp a shmem buffer?

Thanks!

**Comment by kentonv** [comment]
On the sending side, you need to implement a subclass of `capnp::MessageBuilder` which allocates its segments into the shmem buffer. In addition to allocating the segments, it probably has to write some metadata into the shmem buffer communicating the list of segments and where they are located. On the receiving end, you would need a custom subclass of `capnp::MessageReader` that produces segments from the shmem buffer, parsing that same metadata.

Cap'n Proto itself doesn't currently offer an off-the-shelf solution for this, but I don't think it should be all that complicated.

There is also a library called [Flow-IPC](https://github.com/Flow-IPC) which was recently open sourced by Akamai, which implements a featureful shared memory transport that supports Cap'n Proto messages and other things. I haven't vetted this code but it looks interesting.
